src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2247,44 +2247,61 @@ internal void AddModuleInitializerMethod(MethodSymbol method)
             LazyInitializer.EnsureInitialized(ref _moduleInitializerMethods).Add(method);
         }
 
-        private ConcurrentSet<(InterceptsLocationAttributeData, MethodSymbol)>? _interceptions;
+        // NB: the 'Many' case for these dictionary values means there are duplicates. An error is reported for this after binding.
+        private ConcurrentDictionary<(string FilePath, int Line, int Character), OneOrMany<(Location AttributeLocation, MethodSymbol Interceptor)>>? _interceptions;
 
-        internal void AddInterception(InterceptsLocationAttributeData location, MethodSymbol interceptor)
+        internal void AddInterception(string filePath, int line, int character, Location attributeLocation, MethodSymbol interceptor)
         {
             Debug.Assert(!_declarationDiagnosticsFrozen);
-            LazyInitializer.EnsureInitialized(ref _interceptions).Add((location, interceptor));
+
+            var dictionary = LazyInitializer.EnsureInitialized(ref _interceptions);
+            dictionary.AddOrUpdate((filePath, line, character),
+                addValueFactory: static (key, newValue) => OneOrMany.Create(newValue),
+                updateValueFactory: static (key, existingValues, newValue) =>
+                {
+                    // AddInterception can be called when attributes are decoded on a symbol, which can happen for the same symbol concurrently.
+                    // If something else has already added the interceptor denoted by a given `[InterceptsLocation]`, we want to drop it.
```

Comment:
>by a given `[InterceptsLocation]` [](http://example.com/codeflow?start=83&length=33)

Are we reporting an error for duplicate `[InterceptsLocation]` attributes from the same interceptor?

Reply:
Adding a test. I think this area could probably use more testing generally, with interceptors being duplicated across different types, different files, etc.